### PR TITLE
Include flags for unpushed and uncommitted changes, plus deleted flag

### PR DESCRIPTION
**Pull request by Jongmassey**
Forms part of https://github.com/opensafely-core/research-template-docker/issues/70

Record which codespaces have uncommited and/or unpushed changes in order to support reporting on which codespaces users might be at risk of losing data upon codespace deletion.

Once deployed, grafana dashboard will be written to show codespaces near to end of retention period with unpushed/committed changes.

Since the codespaces table is a persistent table (unlike the others) and this PR adds some new columns, we need some sort of way of migrating the database changes. This PR includes a very basic implementation of such a migration, without needing to reach for something more full-featured like Alembic.

### CODE REVIEW

**Comment by Jongmassey** [comment]
Specifically tested that data in the `github_codespaces` table before the addition of these fields persists post the migration, with `NULL` values for these new fields as expected

**Review by iaindillingham — COMMENTED** [review]
Thanks for writing clear commit messages for a concise, and logically-sequenced, set of commits, Jon. I focused my attention on cd4b45707abfb20ecdf2aafe03889dba315f681a, because this is where the complexity lies. I'd like to know more about the complexity introduced by Alembic; a discussion about persistent tables would also be useful.

File: metrics/timescaledb/db.py [hunk 1]
```diff
@@ -143,7 +144,40 @@ def _drop_child_tables(connection, table):
...
+    else:
+        db_table = metadata.tables[table.name]
+        for column in [
+            c for c in table.columns if c.name not in [d.name for d in db_table.columns]
+        ]:
+            connection.execute(text(add_column(table, column, engine)))
+        for column in [
+            d for d in db_table.columns if d.name not in [c.name for c in table.columns]
+        ]:
+            connection.execute(text(drop_column(table, column, engine)))
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply metrics/timescaledb/db.py:191]: > I don't think we should make alter_table, add_column, and drop_column inner functions of _ensure_table. We don't need the closure; or we don't make use of it, because we pass table and engine, which we obtain from connection, to all three. So, if we are going to make them inner functions, then we should make use of the closure.

I agree. The query composition logic was previously a bit more complex but I found I could radically simplify it and thus those functions are pretty redundant. I'll remove them.

>I think the listcomp in a listcomp in a for loop hard to parse. Could we obtain columns to add and columns to drop, and then operate on them?

Yeah, I agree that'd be a lot clearer

**Comment by Jongmassey** [comment]
AIUI this is the basis of the need for persistent Codespaces data in the db (cf the tables which are dropped and recreated and repopulated each run):

The data returned from the GitHub Codespaces API endpoint is fundamentally different from the others which we use for metrics (PRs, Issues, Repos etc.) in that the former are never deleted (repos may be, but rarely) and just have their statuses updated. Therefore it is easy to calculate metrics like "how many PRs were there in each repo at every point in time since its creation?" from scratch each time. However, when Codespaces are deleted their information is no longer returned by the API and so if we dropped all data and recreated we'd only ever know the current picture.

Therefore, we don't use a hypertable for GitHub Codespaces and just upsert the current state of every Codespace rather than calculating any metrics when the task runs.

_(We could change to a strategy of actually using a hypertable and storing the state of each Codespace at a given time - which would enable us to answer questions like "how has the distribution of Codespace's lifespan changed over time?", but would require some rewriting of our existing dashboards and doesn't make the need for a persistent table go away)_

Since we have a table with data already in it we don't want to throw away, any time we make changes to the data from the Codespaces API we store in the table we probably will need some form of migration from the old table schema to the new.

The only way I can see avoiding needing migrations is to store the whole API response json in the table and then use PostgreSQL's json querying functions in the dashboard but this feels _deeply unappealing_.

In the case of this PR, we need to record the flags for a codespace having uncommitted or unpushed data to ascertain which users may be at risk of data loss if their codespace is deleted (e.g. at the end of its idle retention period). We also need to flag which codespaces have already been deleted in order to avoid false positives.

I'd be interested to hear any interesting ideas of how we could do this in a way that avoids the need for migrations.

**Comment by Jongmassey** [comment]
On the assumption that we need a persistent table and therefore migrations here are the options I have considered:

### one off manual table alteration 

through the `psql` console on dokku3 - quick, easy, risky, awkward if we need to restore from backup prior to this change


### Alembic

[Alembic](https://alembic.sqlalchemy.org/en/latest/) is the officially supported current sqlalchemy migrations library.

* reliable and tested
* fully featured
* set up cost
* more dependencies
* [limited timescaledb compatibility](https://stackoverflow.com/questions/75627197/alembic-how-to-create-hypertable) 

I tried working through the tutorial and struggled to get it working without having to write lots of python to create and populate the tables as the initial migration step. At this point I backed out. It seems there is an [alternative strategy](https://medium.com/@cemdurak/alembic-existing-db-a4cf36a33c77) to getting it up and running with an existing database that involves taking a dump of the database and running that as the initial migration step, but I've not tried this.


### roll our own "just enough" migrations

I realised that in this case we already have some mechanics for creating tables via sqlalchemy (`ensure_tables()`) and thought that it wouldn't be too much more effort to extend it to alter tables to add/remove columns (what's needed for this PR). There aren't native `ALTER TABLE` expressions in sqlalchemy so one has to construct these statements as text, and thus there is no guarantee of cross-rdbms compatibility. 

There are still some outstanding issues around my first attempt at implementing this (sqlalchemy's table metadata not updating until the transaction containing the DDL is committed, which means the answer to "drop this table then check if it exists" in a single transaction can be wrong), but even attempting this at all might be not the right approach. Unforeseen future traps, maintainability headaches etc.

**Comment by Jongmassey** [comment]
close for now, let's do this in bennett bot, keep this branch for further discussion re: persistent table problem
